lightning/src/routing/network_graph.rs

```diff
@@ -0,0 +1,1672 @@
+	fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
+		write!(f, "last_update {}, enabled {}, cltv_expiry_delta {}, htlc_minimum_msat {}, fees {:?}", self.last_update, self.enabled, self.cltv_expiry_delta, self.htlc_minimum_msat, self.fees)?;
+		Ok(())
+	}
+}
+
+impl_writeable!(DirectionalChannelInfo, 0, {
+	last_update,
+	enabled,
+	cltv_expiry_delta,
+	htlc_minimum_msat,
+	fees,
+	last_update_message
+});
+
+#[derive(PartialEq)]
+/// Details regarding a channel (both directions)
+pub struct ChannelInfo {
+	/// Protocol features of a channel communicated during its announcement
+	pub features: ChannelFeatures,
+	/// Source node of one of the directions of a channel
+	pub one_src_node: PublicKey,
+	/// Details regarding one of the directions of a channel
+	pub one_to_two: Option<DirectionalChannelInfo>,
+	/// Source node of another direction of a channel
+	pub two_src_node: PublicKey,
+	/// Details regarding another direction of a channel
+	pub two_to_one: Option<DirectionalChannelInfo>,
+	/// An initial announcement of the channel
+	//this is cached here so we can send out it later if required by initial routing sync
+	//keep an eye on this to see if the extra memory is a problem
+	pub announcement_message: Option<msgs::ChannelAnnouncement>,
+}
+
+impl std::fmt::Display for ChannelInfo {
+	fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
+		write!(f, "features: {}, one_src_node: {}, one_to_two: {:?}, two_src_node: {}, two_to_one: {:?}",
+		   log_bytes!(self.features.encode()), log_pubkey!(self.one_src_node), self.one_to_two, log_pubkey!(self.two_src_node), self.two_to_one)?;
+		Ok(())
+	}
+}
+
+impl_writeable!(ChannelInfo, 0, {
+	features,
+	one_src_node,
+	one_to_two,
+	two_src_node,
+	two_to_one,
+	announcement_message
+});
+
+
+/// Fees for routing via a given channel or a node
+#[derive(Eq, PartialEq, Copy, Clone, Debug)]
+pub struct RoutingFees {
+	/// Flat routing fee
+	pub base_msat: u32,
+	/// Liquidity-based routing fee
+	pub proportional_millionths: u32,
+}
+
+impl Readable for RoutingFees{
+	fn read<R: ::std::io::Read>(reader: &mut R) -> Result<RoutingFees, DecodeError> {
+		let base_msat: u32 = Readable::read(reader)?;
+		let proportional_millionths: u32 = Readable::read(reader)?;
+		Ok(RoutingFees {
+			base_msat,
+			proportional_millionths,
+		})
+	}
+}
+
+impl Writeable for RoutingFees {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), ::std::io::Error> {
+		self.base_msat.write(writer)?;
+		self.proportional_millionths.write(writer)?;
+		Ok(())
+	}
+}
+
+#[derive(PartialEq, Debug)]
+/// Information received in the latest node_announcement from this node.
+pub struct NodeAnnouncementInfo {
+	/// Protocol features the node announced support for
+      pub features: NodeFeatures,
+	/// When the last known update to the node state was issued
+      pub last_update: u32,
+	/// Color assigned to the node
+	pub rgb: [u8; 3],
+	/// Moniker assigned to the node
```

Comment:
Some details about this possibly being invalid or malicious (eg control chars) would be good. Probably worth suggesting this not be exposed to users.